apis/apiextensions/v1/composition_transforms.go

```diff
@@ -210,7 +213,7 @@ type StringTransform struct {
 
 	// Convert the type of conversion to Upper/Lower case.
 	// +optional
-	// +kubebuilder:validation:Enum=ToUpper;ToLower
+	// +kubebuilder:validation:Enum=ToUpper;ToLower;ToBase64;FromBase64
```

Comment:
These changes (just the type definitions, not the methods) need to be reflected in the v1alpha1 composition revision types as well. It actually looks like that file is already out-of-date (it doesn't seem to include ToUpper and ToLower for example). I think we'll need to bring it up-to-date in a separate commit so that we can back port that commit without back porting the rest of this feature.

Reply:
OK, so two commits - one to add the ToUpper/ToLower which can be backported, and a second to add the ToBase64/FromBase64 that will not be backported?

I'm not entirely sure what the scope of the changes should be, but this is what I came up with:

```
diff --git a/apis/apiextensions/v1alpha1/revision_types.go b/apis/apiextensions/v1alpha1/revision_types.go
index bbddaf5c..a7d3018f 100644
--- a/apis/apiextensions/v1alpha1/revision_types.go
+++ b/apis/apiextensions/v1alpha1/revision_types.go
@@ -408,12 +408,34 @@ func (m *MapTransform) Resolve(input interface{}) (interface{}, error) {
        }
 }
 
+// StringTransformType is type of the string transform function to be executed fmt/convert.
+type StringTransformType string
+
+// StringConversionType is the type of string conversion, ToUpper/ToLower/ToBase64/FromBase64
+type StringConversionType string
+
 // A StringTransform returns a string given the supplied input.
 type StringTransform struct {
+
+       // Type of the string transform to be run.
+       // +optional
+       // +kubebuilder:validation:Enum=Format;Convert;TrimPrefix;TrimSuffix
+       // +kubebuilder:default=Format
+       Type StringTransformType `json:"type,omitempty"`
+
        // Format the input using a Go format string. See
        // https://golang.org/pkg/fmt/ for details.
        // +immutable
-       Format string `json:"fmt"`
+       Format string `json:"fmt,omitempty"`
+
+       // Convert the type of conversion to Upper/Lower case.
+       // +optional
+       // +kubebuilder:validation:Enum=ToUpper;ToLower
+       Convert *StringConversionType `json:"convert,omitempty"`
+
+       // Trim the prefix or suffix from the input
+       // +optional
+       Trim *string `json:"trim,omitempty"`
 }
 
 // A ConvertTransform converts the input into a new object whose type is supplied.
```

Is that right?   I wasn't sure if Format should still be "immutable", but making it "optional" changed the type to *string.

Thanks!

Reply:
> I assume it needs two commits - one to add the ToUpper/ToLower which can be backported, and a second to add the ToBase64/FromBase64 that will not be backported?

Yes exactly, and I appreciate you tackling that! The diff you posted looks good to me.

Reply:
Thanks!  These changes are pushed too.